File: 0392_is_subsequence/python/solution_followup.py
from bisect import bisect_left
from collections import defaultdict


class SubsequenceChecker:
    """Preprocess t once, then check many s strings efficiently."""

    def __init__(self, t: str) -> None:
        self.index_map: dict[str, list[int]] = defaultdict(list)
        for i, ch in enumerate(t):
            self.index_map[ch].append(i)

    def is_subsequence(self, s: str) -> bool:
        pos = 0  # current position in t (must match at or after this index)
        for ch in s:
            indices = self.index_map.get(ch)
            if indices is None:
                return False
            idx = bisect_left(indices, pos)
            if idx == len(indices):
                return False
            pos = indices[idx] + 1
        return True

File: 0392_is_subsequence/go/solution_followup.go
package is_subsequence

import "sort"

type SubsequenceChecker struct {
	indexMap [26][]int
}

func NewSubsequenceChecker(t string) *SubsequenceChecker {
	sc := &SubsequenceChecker{}
	for i, ch := range t {
		sc.indexMap[ch-'a'] = append(sc.indexMap[ch-'a'], i)
	}
	return sc
}

func (sc *SubsequenceChecker) IsSubsequence(s string) bool {
	pos := 0
	for _, ch := range s {
		indices := sc.indexMap[ch-'a']
		idx := sort.SearchInts(indices, pos)
		if idx == len(indices) {
			return false
		}
		pos = indices[idx] + 1
	}
	return true
}
